skill_predictor.py

```diff
@@ -0,0 +1,210 @@
+import kenlm
+import numpy as np
+
+import os
+import heapq
+import json
+import collections
+import random
+import datetime
+import itertools
+
+
+class SkillPredictor(object):
+    def __init__(self):
+        self.data = []
+
+        # make sure skills.json is in CWD
+        with open('skills.json') as f:
+            for line in f:
+                self.data.append(json.loads(line))
+
+        # dict mapping each user to the skills it has at each particular time stamp
+        self.user_skills = {}  # str -> int -> str
+        self.all_skills = set()
+        for user in self.data:
+            curr_user_skills = collections.defaultdict(set)
+            curr_skills = set()
+            curr_time = user['skills'][0]['src']['at']
+            for s in user['skills']:
+                time_stamp = s['src']['at']
+                curr_skill = s['val']
+                curr_skill = curr_skill.replace(" ", "")
+                curr_skill = curr_skill.lower()
+                curr_user_skills[time_stamp].add(curr_skill)
+                self.all_skills.add(curr_skill)
+            self.user_skills[user['id']] = curr_user_skills
+
+        self.mle_bagsize = {}
+        self.skill_acquisition_times = None
+        self.model = None
+
+    def get_bag_size_statistics(self):
+        """
+        Finds the average bag size and number of bags for a dataset.
+        """
+        avg_bag_size = 0
+        num_bags = 0
+        for user, bags in self.user_skills.items():
+            if len(bags) <= 1:
+                continue
+            for bag in bags.values():
+                if len(bag) > 2:
+                    avg_bag_size += len(bag)
+                    num_bags += 1
+        return 'Average bag size: ' + str(avg_bag_size / num_bags) + '; Number bags: ' + str(num_bags)
+
+    def _generate_skill_acquisition_times(self):
+        self.skill_acquisition_times = collections.defaultdict(list)
+        for user, time_stamps in self.user_skills.items():
+            entries = list(time_stamps.keys())
+            if len(entries) <= 1:
+                continue
+            for i in range(1, len(entries)):
+                curr_skills = time_stamps[entries[i]]
+                prev_skills = time_stamps[entries[i-1]]
+                curr_day = datetime.datetime.fromtimestamp(entries[i] / 1000)
+                prev_day = datetime.datetime.fromtimestamp(entries[i-1] / 1000)
```

Comment:
'day' is a littl misleading here, because ter might be a much onger gap between entries. 'period' would be a better descriptor.